File: leetcode/sqrtx.go
// 69. Sqrt(x)
// difficulty: easy
// https://leetcode.com/problems/sqrtx/

package leetcode

func mySqrt(x int) int {
	return binarySearchSqrt(x)
	// return bruteForceSqrt(x)
}

func binarySearchSqrt(x int) int {
	left, right, v := 1, x, x/2

	if x == 0 {
		return 0
	}

	for v >= 1 {
		if v*v > x {
			right = v
			v /= 2
		} else if v*v < x {
			left = max(v, left)
			v = (left + right) / 2
			if left == v {
				return left
			}
		} else {
			return v
		}
	}
	return left
}

func max(x int, y int) int {
	if x > y {
		return x
	} else {
		return y
	}
}

func bruteForceSqrt(x int) int {
	v := 0
	for v*v <= x {
		v++
	}
	return v - 1
}

File: leetcode/sqrtx.py
# 69. Sqrt(x)
# difficulty: easy

import pytest


class Solution:
    def mySqrt(self, x: int) -> int:
        return binary_search_sqrt(x)


def binary_search_sqrt(x):
    left, right, v = 1, x, x // 2

    if x == 0:
        return 0

    while v >= 1:
        if v * v > x:
            right = v
            v //= 2
        elif v * v < x:
            left = max(left, v)
            v = (left + right) // 2
            if left == v:
                return v
        else:
            return v
    return left


@pytest.mark.parametrize('x, expected', [
    (0, 0),
    (1, 1),
    (2, 1),
    (3, 1),
    (4, 2),
    (8, 2),
    (9, 3),
    (10, 3),
    (15, 3),
    (16, 4),
    (17, 4),
    (19, 4),
    (1000, 31),
    (1234, 35),
    (12345, 111),
    (777777, 881),
])
def test(x, expected):
    s = Solution()
    assert s.mySqrt(x) == expected


if __name__ == '__main__':
    pytest.main(['-v', __file__])
